src/MELT/TestSink.cs

```diff
@@ -51,20 +64,75 @@ public void Write(WriteContext context)
             MessageLogged?.Invoke(context);
         }
 
-        public void BeginScope(BeginScopeContext context)
+        public IDisposable BeginScope(BeginScopeContext context)
         {
+            TestScope testScope;
+            lock (_scopeLock)
+            {
+                var oldScope = _currentScope;
+                testScope = new TestScope(this, oldScope);
+
+                //In principle, using Immutable collections here would prevent copying the collections
+                //But the number of scopes is low enough that it probably isn't worth the extra dependency
+                var newScopeData = oldScope.ScopeData.Prepend(new BeginScope(context)).ToList();
+
+                var newActiveScopes = oldScope.ActiveScopes.Append(testScope).ToList();
+                _currentScope = (newScopeData,newActiveScopes);
+            }
+
+
```

Comment:
nit: extra blank line, like in few other places

Reply:
Sorry about this, formatting is not my strong point. I've gone through and fixed it everywhere I think.